packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,180 @@
+import { NodePHP } from '@php-wasm/node';
+
+const Fatal = Symbol('Fatal');
+const Spawn = Symbol('Spawn');
+const Reap = Symbol('Reap');
+
+let childCount = 0;
+
+export class PoolInfo {
+	id = childCount++;
+	requests = 0;
+	started = Date.now();
+	active = false;
+}
+
+/**
+ * Maintains and refreshes a list of php instances
+ * such that each one will only be fed X number of requests
+ * before being discarded and replaced.
+ */
+export class Pool {
+	instances = new Map(); // php => PoolInfo
+
+	spawner: () => Promise<any>; // Callback to create new instances.
+	maxRequests: number; // Max requests to feed each instance
+	maxJobs: number; // Max number of instances to maintain at once.
+
+	notifiers = new Map(); // Inverted promises to notify async code of backlogged item processed.
+	running = new Set(); // Set of busy PHP instances.
+	backlog = []; // Set of request callbacks waiting to be run.
+
+	constructor({
+		spawner = async (): Promise<any> => {},
+		maxRequests = 2000,
+		maxJobs = 5,
+	} = {}) {
+		Object.assign(this, { spawner, maxRequests, maxJobs });
+		this[Reap]();
+		this[Spawn]();
+	}
+
+	/**
+	 * Find the next available idle instance.
+	 */
+	getIdleInstance() {
+		const sorted = [...this.instances].sort(
+			(a, b) => a[1].requests - b[1].requests
+		);
+
+		for (const [instance, info] of sorted) {
+			if (this.running.has(instance)) {
+				continue;
+			}
+
+			if (!info.active) {
+				continue;
+			}
+			return instance;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Queue up a callback that will make a request when an
+	 * instance becomes idle.
+	 */
+	async enqueue(item: (php: NodePHP) => Promise<any>) {
+		const idleInstance = this.getIdleInstance();
+
+		if (!idleInstance) {
+			// Defer the callback if we don't have an idle instance available.
+			this.backlog.push(item);
+
+			// Split a promise open so it can be accepted or
+			// rejected later when the item is processed.
+			const notifier = new Promise((accept, reject) =>
+				this.notifiers.set(item, [accept, reject])
+			);
+
+			return notifier;
+		} else {
+			// If we've got an instance available, run the provided callback.
+
+			// When the provided callback completes, check to see if
+			// any more requests have been added to the pool
+			const onCompleted = async () => {
+				this.running.delete(idleInstance);
+
+				this[Reap]();
+				this[Spawn]();
+
+				if (!this.backlog.length) {
+					return;
+				}
+
+				const idleInstanceNext = this.getIdleInstance();
+
+				const next = this.backlog.shift();
+				const info = this.instances.get(idleInstanceNext);
```

Comment:
couldn't `ideInstanceNext` here be false too?

Reply:
While this function is kicked off by the resolution of a previous promise that SHOULD free up the instance its using, I could see how it would be possible for a FALSE to come in here if we get a request at JUST the wrong millisecond.

This is a little complicated because we're actually managing TWO pools here, one of backlogged requests and one of idle/running PHP instances, but I think I can come up with something robust to that case.